### ISSUE TITLE
suggest replacing if_let and ifs with matches and ifs

### ISSUE THREAD

**Issue by J-ZhengLi** [rust-lang/rust-clippy#10513]
### What it does

suggest using `matches` macro with guards over syntax like:

```rust
if let Enum(x) = enum {
    if x.blahblah() {
        // ...
    }
}
```

**when `x` is only used for checking purpose.**

### Lint Name

if_let_with_if?

### Category

~~style~~

restriction

### Advantage

This seems cleaner in my opinion

### Drawbacks

might have some FP.

### Example

```rust
let maybe_header = Some("# Title");

if let Some(text) = maybe_header {
    if text.starts_with("#") {
        println!("header detected");
    }
}
```

Could be written as:

```rust
if matches!(maybe_header, Some(text) if text.starts_with("#")) {
    println!("header detected");
}
```

**Comment by blyxyas** [comment]
I think that this lint should be category "restriction" because style lints [are Warn level] and it is really situation and author dependant to decide if an "if let" statement should be reworded to use `matches!`

**Comment by J-ZhengLi** [comment]
Seems hard, but I gotta get a taste of my own medicine... 
also, I might name it `manual_matches`
@rustbot claim

**Comment by J-ZhengLi** [comment]
nvm, there's a `match_like_matches_macro` does similar things, closing now

**Comment by J-ZhengLi** [comment]
Actually... on a second thought, `match_like_matches_macro` only checks if-let/match phrases that produces bool, which could use some improvement imo...

So... reopening now...
